src/Tmds.Ssh/GssapiWithMicCredential.cs

```diff
@@ -0,0 +1,282 @@
+// This file is part of Tmds.Ssh which is released under MIT.
+// See file LICENSE for full license details.
+
+using System;
+using System.Buffers;
+using System.Linq;
+using System.Net;
+using System.Net.Security;
+using System.Reflection;
+using System.Security.Principal;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Tmds.Ssh;
+
+public sealed class GssapiWithMicCredential : Credential
+{
+    // gssapi-with-mic is defined in https://datatracker.ietf.org/doc/html/rfc4462
+    private const string AUTH_METHOD = "gssapi-with-mic";
+
+    // Kerberos - 1.2.840.113554.1.2.2
+    private static readonly byte[] KRB5_OID = [ 0x06, 0x09, 0x2A, 0x86, 0x48, 0x86, 0xF7, 0x12, 0x01, 0x02, 0x02 ];
+
+    private readonly Func<string?> _getPassword;
+    private readonly bool _delegateCredential;
+    private readonly string? _serviceName;
+
+#if NET8_0
+    private delegate void GetMICDelegate(ReadOnlySpan<byte> data, IBufferWriter<byte> writer);
+
+    private static readonly Lazy<MethodInfo?> _getMicMethInfo = new Lazy<MethodInfo?>(() =>
+        typeof(NegotiateAuthentication).GetMethod("GetMIC", BindingFlags.NonPublic | BindingFlags.Instance));
+
+    private MethodInfo GetMICMethod { get; }
+#endif
+
+    /// <summary>
+    /// Create a GSSAPI user authentication context with password string.
+    /// </summary>
+    /// <param name="password">The password to use for the GSSAPI username. Set to null to use a cached credential.</param>
+    /// <param name="delegateCredential">Request delegation on the GSSAPI context.</param>
+    /// <param name="serviceName">Override the service principal name (SPN), default it to use the connection hostname</param>
+    public GssapiWithMicCredential(string? password = null, bool delegateCredential = false, string? serviceName = null) : this(() => password, delegateCredential, serviceName)
```

Comment:
suggested rename: `delegateCredential` to `forwardCredentials`.

Reply:
suggested rename: `serviceName` to `serverIdentity`.

Reply:
Up to you, I went with `delegateCredential` because the GSSAPI/SSPI flag has delegation in the name. The OpenSSH option for this is also `GSSAPIDelegateCredentials` to there is some consistency there (I can make it plural if you like).

For `serviceName` I settled on that because it refers to the Service Principal Name (SPN) which is a common term used in Kerberos. Happy to rename either though if you prefer that.

Reply:
> Up to you, I went with delegateCredential because the GSSAPI/SSPI flag has delegation in the name. The OpenSSH option for this is also GSSAPIDelegateCredentials to there is some consistency there (I can make it plural if you like).

Let's keep the OpenSSH based name.

> For serviceName I settled on that because it refers to the Service Principal Name (SPN) which is a common term used in Kerberos. Happy to rename either though if you prefer that.

If this a common Kerberos term. Let's also keep it.